### PR TITLE
Moves 2D laser options into 2D trajectory builder.

### PR DESCRIPTION
**Pull request by damonkohler**

### CODE REVIEW

File: configuration_files/trajectory_builder_2d.lua [hunk 1]
```diff
@@ -14,6 +14,9 @@
 
 TRAJECTORY_BUILDER_2D = {
   use_imu_data = true,
+  laser_min_range = 0.,
+  laser_max_range = 30.,
+  laser_missing_echo_ray_length = 5.,
```
Comment (wohe) [inline configuration_files/trajectory_builder_2d.lua]: Consistency: Move below 'laser_max_z'?

File: cartographer/mapping_2d/local_trajectory_builder.cc [hunk 2]
```diff
@@ -73,8 +79,19 @@ const Submaps* LocalTrajectoryBuilder::submaps() const { return &submaps_; }
 sensor::LaserFan LocalTrajectoryBuilder::BuildCroppedLaserFan(
     const transform::Rigid3f& tracking_to_tracking_2d,
     const sensor::LaserFan& laser_fan) const {
+  LaserFan fan{laser_fan.origin, {}, {}, laser_fan.reflectivities};
```
Comment (wohe) [inline cartographer/mapping_2d/local_trajectory_builder.cc]: Should reflectivities be retained here? And if so: Do we need to drop/reorder reflectivities when elements are dropped from 'returns'?

File: cartographer/mapping_2d/local_trajectory_builder.cc [hunk 3]
```diff
@@ -73,8 +79,19 @@ const Submaps* LocalTrajectoryBuilder::submaps() const { return &submaps_; }
...
+  for (const Eigen::Vector3f& return_ : laser_fan.returns) {
+    const float range = (return_ - laser_fan.origin).norm();
+    if (options_.laser_min_range() <= range &&
+        range <= options_.laser_max_range()) {
+      fan.returns.push_back(return_);
+    }
+  }
+  for (const Eigen::Vector3f& miss : laser_fan.misses) {
+    fan.misses.push_back(options_.laser_missing_echo_ray_length);
```
Comment (wohe) [inline cartographer/mapping_2d/local_trajectory_builder.cc]: This seems strange: Adding a scalar value to the point cloud of misses? "options_.laser_missing_echo_ray_scale * miss"?

File: cartographer/mapping_2d/local_trajectory_builder.cc [hunk 4]
```diff
@@ -73,8 +79,19 @@ const Submaps* LocalTrajectoryBuilder::submaps() const { return &submaps_; }
 sensor::LaserFan LocalTrajectoryBuilder::BuildCroppedLaserFan(
     const transform::Rigid3f& tracking_to_tracking_2d,
     const sensor::LaserFan& laser_fan) const {
+  LaserFan fan{laser_fan.origin, {}, {}, laser_fan.reflectivities};
+  for (const Eigen::Vector3f& return_ : laser_fan.returns) {
+    const float range = (return_ - laser_fan.origin).norm();
+    if (options_.laser_min_range() <= range &&
+        range <= options_.laser_max_range()) {
+      fan.returns.push_back(return_);
```
Comment (wohe) [inline cartographer/mapping_2d/local_trajectory_builder.cc]: I'm not sure I can follow here: Returns outside the range are dropped completely, but misses are retained? Should this code implement the logic currently in sensor::ToLaserFan() and that function be removed/moved to cartographer_ros?

File: cartographer/mapping_2d/local_trajectory_builder.cc (hunk 2 above)
Comment (damonkohler) [inline cartographer/mapping_2d/local_trajectory_builder.cc]: No, thought I had removed that. Fixed.

File: configuration_files/trajectory_builder_2d.lua (hunk 1 above)
Comment (damonkohler) [inline configuration_files/trajectory_builder_2d.lua]: Done.

File: cartographer/mapping_2d/local_trajectory_builder.cc (hunk 4 above)
Comment (damonkohler) [inline cartographer/mapping_2d/local_trajectory_builder.cc]: This was a mess, sorry for that. Should be reaonsable now.

File: cartographer/mapping_2d/local_trajectory_builder.cc (hunk 3 above)
Comment (damonkohler) [inline cartographer/mapping_2d/local_trajectory_builder.cc]: Done.

File: cartographer/sensor/laser.h [hunk 5]
```diff
@@ -38,12 +38,10 @@ struct LaserFan {
   std::vector<uint8> reflectivities;
 };
 
-// Builds a LaserFan from 'proto' and separates any beams with ranges outside
-// the range ['min_range', 'max_range']. Beams beyond 'max_range' are inserted
-// into the 'misses' point cloud with length 'missing_echo_ray_length'. The
-// points in both clouds are stored in scan order.
+// Builds a LaserFan from 'proto' and drops any beams with ranges outside
+// the range ['min_range', 'max_range'].
```
Comment (wohe) [inline cartographer/sensor/laser.h]: Add a comment that this is only to drop invalid data and handling short/long returns and missing beams is expected to be handled later?

File: cartographer/mapping_2d/proto/local_trajectory_builder_options.proto [hunk 6]
```diff
@@ -24,10 +24,15 @@ import "cartographer/mapping_2d/scan_matching/proto/ceres_scan_matcher_options.p
...
 
 message LocalTrajectoryBuilderOptions {
-  // Cropping parameters for the laser fan.
+  // Laser returns outside these ranges will be dropped.
+  optional float laser_min_range = 14;
+  optional float laser_max_range = 15;
   optional float laser_min_z = 1;
   optional float laser_max_z = 2;
 
+  // Missing laser returns will be inserted with this length as empty space.
```
Comment (wohe) [inline cartographer/mapping_2d/proto/local_trajectory_builder_options.proto]: Missing -> Returns outside of laser_max_range?

Comment (damonkohler) [inline cartographer/mapping_2d/proto/local_trajectory_builder_options.proto]: Done.

File: cartographer/sensor/laser.h (hunk 5 above)
Comment (damonkohler) [inline cartographer/sensor/laser.h]: Done.

File: cartographer/mapping_2d/local_trajectory_builder.cc [hunk 7]
```diff
@@ -70,17 +76,31 @@ LocalTrajectoryBuilder::~LocalTrajectoryBuilder() {}
...
+  // maximum range into misses.
+  sensor::LaserFan returns_and_misses{laser_fan.origin, {}, {}, {}};
+  for (const Eigen::Vector3f& return_ : laser_fan.returns) {
+    const float range = (return_ - laser_fan.origin).norm();
+    if (range >= options_.laser_min_range()) {
+        if (range <= options_.laser_max_range()) {
+          returns_and_misses.returns.push_back(return_);
+        } else {
+          returns_and_misses.misses.push_back(
+              options_.laser_missing_echo_ray_length() * return_.normalized());
```
Comment (damonkohler) [inline cartographer/mapping_2d/local_trajectory_builder.cc]: Done.

**Pull request merged**